### PR TITLE
Implement OpenID Connect for authentication

### PR DESCRIPTION
**Pull request by mdp**
Implements the OpenID Connect standard for SSO Authentication
which allows users to use any OpenID Connect provider for authentication
not just a predefined list of providers

User can add OIDC config settings to firezone.rb which will then
populate the AUTH_OIDC environment variable as a JSON string.

FZ_HTTP will use this environment variable to create
provider(s) to authenticate against

Additional notes:
- Updates .env.sample to include an example of a 'stringified' JSON
environment variable for setting AUTH_OIDC in development
- Add dep for 'openid_connect' and test dep for 'mox'

### CODE REVIEW

File: apps/fz_http/test/fz_http_web/controllers/session_controller_test.exs [hunk 1]
```diff
@@ -63,6 +74,45 @@ defmodule FzHttpWeb.AuthControllerTest do
     end
   end
 
+  #   test "signing in from OIDC callback", %{unauthed_conn: conn} do
```
Comment (jamilbk) [inline apps/fz_http/test/fz_http_web/controllers/session_controller_test.exs]: Should this comment be removed or reworded?

File: apps/fz_http/lib/fz_http_web/user_from_auth.ex [hunk 2]
```diff
@@ -24,4 +24,11 @@ defmodule FzHttpWeb.UserFromAuth do
       user -> {:ok, user}
     end
   end
+
+  def find_or_create(_provider, %{"email" => email, "sub" => _sub, "email_verified" => true}) do
```
Comment (jamilbk) [inline apps/fz_http/lib/fz_http_web/user_from_auth.ex:28]: Thanks for the catch here requiring the `email_verified` claim!

File: apps/fz_http/lib/fz_http_web/controllers/auth_controller.ex [hunk 3]
```diff
@@ -48,6 +49,48 @@ defmodule FzHttpWeb.AuthController do
     end
   end
 
+  def callback(conn, params) do
+    %{"provider" => provider_key} = params
+    openid_connect = Application.fetch_env!(:fz_http, :openid_connect)
+
+    atomize = fn key ->
+      try do
+        {:ok, String.to_existing_atom(key)}
```
Comment (jamilbk) [inline apps/fz_http/lib/fz_http_web/controllers/auth_controller.ex:58]: 👍

File: apps/fz_http/lib/fz_http_web/controllers/auth_controller.ex [hunk 4]
```diff
@@ -11,7 +12,7 @@ defmodule FzHttpWeb.AuthController do
   # Uncomment when Helpers.callback_url/1 is fixed
   # alias Ueberauth.Strategy.Helpers
 
-  plug Ueberauth
+  plug(Ueberauth)
```
Comment (jamilbk) [inline apps/fz_http/lib/fz_http_web/controllers/auth_controller.ex]: (nitpick) Wonder if we should change all top-level `plug` calls to include parens like this? Curious to hear @princemaple's thoughts. I've only seen module-level macro functions used without parens but don't feel strongly about it.

File: apps/fz_http/test/fz_http_web/controllers/session_controller_test.exs (hunk 1 above)
Comment (mdp) [reply apps/fz_http/test/fz_http_web/controllers/session_controller_test.exs]: Sorry, totally missed this. Amended and forced-pushed.

File: apps/fz_http/lib/fz_http_web/controllers/auth_controller.ex (hunk 4 above)
Comment (mdp) [reply apps/fz_http/lib/fz_http_web/controllers/auth_controller.ex]: I don't remember doing this manually, but maybe a formatting plugin did this without my knowledge. Reverting to convention.

**Review by jamilbk — APPROVED** [review]
Great work! Thanks for implementing this @mdp

File: apps/fz_http/lib/fz_http_web/controllers/auth_controller.ex (hunk 4 above)
Comment (princemaple) [reply apps/fz_http/lib/fz_http_web/controllers/auth_controller.ex]: It's when dialyzer doing its thing and elixir LS doesn't load the formatter config in time, but is ready to respond to formatting request... It messes up.

I see them and discard them while staging changes.

**Pull request merged**